nixos/modules/services/mail/dovecot.nix

```
@@ -272,22 +206,215 @@ let
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "sieveScripts" ]
+        [ "services" "dovecot2" "sieve" "scripts" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "configVersion" ]
+        [ "services" "dovecot2" "settings" "dovecot_config_version" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "storageVersion" ]
+        [ "services" "dovecot2" "settings" "dovecot_storage_version" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "mailUser" ]
+        [ "services" "dovecot2" "settings" "mail_uid" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "mailGroup" ]
+        [ "services" "dovecot2" "settings" "mail_gid" ]
+      )
+      (mkRenamedOptionModule
+        [ "services" "dovecot2" "protocols" ]
+        [ "services" "dovecot2" "settings" "protocols" ]
+      )
+    ]
+    ++ (
+      let
+        basePath = [
+          "services"
+          "dovecot2"
+        ];
+        mkRemovedOptions =
+          list:
+          map (
+            name: mkRemovedOptionModule (basePath ++ name) "Please use services.dovecot2.settings instead."
+          ) list;
+      in
+      mkRemovedOptions [
+        [ "pluginSettings" ]
+        [ "enableQuota" ]
+        [ "quotaPort" ]
+        [ "quotaGlobalPerUser" ]
+      ]
+    );
 
   options.services.dovecot2 = {
     enable = mkEnableOption "the dovecot 2.x POP3/IMAP server";
 
+    package = mkPackageOption pkgs "dovecot" {
+      default = [
+        (if versionAtLeast config.system.stateVersion "25.11" then pkgs.dovecot else pkgs.dovecot_2_3)
+      ];
+    };
+
+    settings = mkOption {
+      default = { };
+      type =
+        let
+          inherit (lib.types)
+            attrsOf
+            bool
+            int
+            listOf
+            nonEmptyStr
+            nullOr
+            oneOf
+            str
+            submodule
+            ;
+          inherit (lib) elemAt;
+          inherit (lib.lists) last length sublist;
+
+          # Override submodule to pass the parent name to the submodule name arg
+          # if the current name is a list (starting with '[definition ') so that
+          # it can be parsed by the submodule
+          submodule' =
+            f:
+            let
+              orig = submodule f;
+              merge =
+                loc: defs:
+                let
+                  loc' =
+                    if builtins.match ''\[definition .*'' (last loc) != null then
+                      sublist 0 (length loc - 1) loc
+                    else
+                      loc;
+                in
+                orig.merge loc' defs;
+            in
+            orig // { inherit merge; };
+
+          section = submodule' (
+            { name, ... }:
+            let
+              # splits name on the first space
+              splits = builtins.match "([^ ]+) (.+)" name;
```

```suggestion
              splits = builtins.match "([^ ]+) +(.+)" name;
```

see https://github.com/NixOS/nixpkgs/pull/410912#discussion_r2107669499. On second thought, not sure if we *should* do this since it makes keys `a b` and `a  b` map to the same section in the config.

For the current code to not do the same though in the end, it would have to implement escaping in the config printer though ([function settings_section_escape in settings-parser.c](https://github.com/dovecot/core/blob/42864e4182c7bfe2931fdf888f3cb4ba77953dee/src/lib-settings/settings-parser.c#L1153)). Maybe a good idea to implement this because sometimes the section name is interpreted so you might want some of those special chars in the name.

---

If it uses backslash escapes, we should be able to use `lib.strings.escape`.

E.g. `escape [ "/" "+" "." ] name`

---

do you mean something like this?

```
nix-repl> :p builtins.replaceStrings ["=" ''\'' " " ","] [''\e'' ''\\'' ''\_'' ''\+''] ''test \ s.ect_ion , na=me''
test\_\\\_s.ect_ion\_\+\_na\eme
```